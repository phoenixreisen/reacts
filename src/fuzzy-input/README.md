# Phoenix Fuzzy-Suche

**Ein Eingabefeld, das während des Tippens bereits Suchergebnisse anzeigt.** Die Komponente stellt nur die Implementierung, 
die Styles kommen wie immer aus dem Design-System.

Es gibt zwei Varianten:

- **Autocomplete für komplette Eingabe.** Entsprechend der Eingabe ins Textfeld werden Vorschläge abgerufen und angezeigt. Wird ein Vorschlag ausgewählt, wird die komplette Eingabe durch diesen Vorschlag ersetzt. Siehe Demo.

- **Autocomplete für Platzhalter.** Tippt man innerhalb eines Textes ein definiertes Prefix, werden entsprechende Platzhalter abgerufen und angezeigt. Wird ein Platzhaler ausgewählt, wird der Prefix (sowie das direkt anschließend Getippte bis zum nächsten Leerzeichen) durch diesen Platzhalter (samt Prefix+Suffix) ersetzt. Der Rest des Textes bleibt bestehen. Siehe Demo.

Die Komponente ist Teil des [Phoenix Reisen Design-Systems](https://design-system.phoenixreisen.net).

## Demo

https://phoenixreisen.github.io/reacts/fuzzy-input/

## Installation

[React](https://react.dev/) wird benötigt.

```bash
npm install --save @phoenixreisen/reacts
```

## Anwendung

#### Parameter / Props

```ts
type Attrs = {
    id?: string,
    label?: string,
    value?: string,
    valid?: boolean,
    pattern?: RegExp,
    warnmsg?: string,
    errormsg?: string,
    disabled?: boolean,
    readonly?: boolean,
    logerror?: boolean,
    maxLength?: number,
    minLength?: number,
    throttling?: number,
    withButton?: boolean,
    placeholder?: string,
    clearAfterLoad?: boolean,
    inText?: {
        prefix: string,
        suffix: string,
    },
    onLoad: (name: string) => Promise<unknown>,
    onQuery: (input?: string) => Promise<Array<string>>
    onInput: (input: string) => void,
    onBlur?: () => void,
}
```

#### Aufruf

```ts
/** Autocomplete gesamter Eingabe (siehe Demo) */

import FuzzyInput from '@phoenixreisen/reacts/fuzzy-input';
import React from 'react';

// Entweder JSX
<FuzzyInput
    id={'example-1'}
    label={'Vorlagensuche'}
    pattern={new RegExp(/[a-Z]/)}
    warnmsg={'Ungueltige Eingabe'}
    errormsg={'Huch, ein Fehler ist aufgetreten.'}
    oninput={(value) => someState.value = value}
    query={(needle: string) => Promise.resolve(console.log('search it'))}
    load={(choice: string) => Promise.resolve(console.log('get it'))}
    value={() => someState.value}
/>

//oder Hyperscript bzw. JS
m(FuzzyInput, {
    id: 'example-1',
    label: 'Vorlagensuche',
    pattern: new RegExp(/[0-9]/),
    warnmsg: 'Ungültige Eingabe',
    errormsg: 'Es ist ein Fehler aufgetreten.',
    oninput: (value) => someState.value = value}
    load: (name: string) => Promise.resolve(console.log('search it')),
    query: (input: string) => Promise.resolve(console.log('get it')),
    value: () => someState.value}
});
```

```ts
/** Autocomplete für Platzhalter mit Button (siehe Demo) */

import FuzzyInput from '@phoenixreisen/reacts/fuzzy-input';
import React from 'react';

<FuzzyInput
    inText={{
        prefix: '{{',
        suffix: '}}'
    }}
    minLength={1}
    id={'example-2'}
    withButton={true}
    label={'Irgendwas'}
    pattern={new RegExp(/[{a-zA-Z]/)}
    warnmsg={'Ungueltige Eingabe'}
    value={() => someState.value}
    oninput={(value) => someState.value = value}
    errormsg={'Huch, ein Fehler ist aufgetreten.'}
    query={(needle: string) => query2(needle)}
    load={(choice: string) => load2(choice)}
/>
```

## Demo für Github Page

Mit folgendem Befehl werden alle Demos in den Ordner [../../docs](../../docs) gebaut. Dieser Ordner wird von Github zur Bereitstellung der Demos bzw. Github Pages genutzt.

Möchte man zum Entwickeln nur bestimmte Demos bauen, kann man in der [rollup.config.js](../../rollup.config.js) nicht benötigte Demos auskommentieren, sollte sie vor dem `push` aber wieder einkommentieren.

```bash
npm run compile:demos
```

## Test

```bash
[npm install]
npm run test
```

## Deployment

Deployed & published wird immer die ganze Sammlung. [Siehe hier](../../README.md).