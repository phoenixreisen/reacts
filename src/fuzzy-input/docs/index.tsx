import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import { createRoot } from 'react-dom/client';
import { Demo } from './fuzzy.demo';
import React from 'react';

export const Root = () => {
    return (
        <Demo />
    );
};

createRoot(document.querySelector('.example-app')).render(
    <Router>
        <Routes>
            <Route path="*" element={<Root />} />
        </Routes>
    </Router>
);
