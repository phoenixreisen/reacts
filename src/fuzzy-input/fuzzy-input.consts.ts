//--- Konstanten, Variablen & Objekte -----

/** Defaults */
export const ID = 'fuzzy-input'; // el id

export const THROTTLING = 500; // ms
export const MAXLENGTH = 100; // chars
export const MINLENGTH = 3; // chars
