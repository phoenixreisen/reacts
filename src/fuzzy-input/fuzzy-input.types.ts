//--- View Types -----

export type Props = {
    inText?: {
        prefix: string,
        suffix: string,
    },
    id?: string,
    value: string,
    label?: string,
    valid?: boolean,
    pattern?: RegExp,
    warnmsg?: string,
    errormsg?: string,
    disabled?: boolean,
    readonly?: boolean,
    logerror?: boolean,
    maxLength?: number,
    minLength?: number,
    throttling?: number,
    withButton?: boolean,
    placeholder?: string,
    clearAfterLoad?: boolean,
    // Event-Funktionen
    onInput: (input: string) => void,
    onBlur?: (param?: unknown) => void,
    onLoad: (name: string) => Promise<unknown>,
    onQuery: (input?: string) => Promise<Array<string>>,
    onFocusCallback?: (input: string, state: State, props: Props) => void,
}

export type State = {
    value: string,
    valid: boolean,
    loading: boolean,
    error: Error|null,
    events: Events | null,
}

export type Events = {
    ESCAPE: ((e: KeyboardEvent) => void) | null,
    ARROW: ((e: KeyboardEvent) => void) | null
}