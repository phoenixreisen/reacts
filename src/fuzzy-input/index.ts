import FuzzyInput from './fuzzy-input.r';

export {
    Props,
    State,
    Events
} from './fuzzy-input.types';

export {
    ID,
    MAXLENGTH,
    THROTTLING,
    MINLENGTH,
} from './fuzzy-input.consts';

export {
    isValid,
    isReady,
    findMatch,
} from './fuzzy-input.fns';

export {
    FuzzyInput
} from './fuzzy-input.r';

export default FuzzyInput;