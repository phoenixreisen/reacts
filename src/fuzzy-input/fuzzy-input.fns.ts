import { State, Props } from './fuzzy-input.types';
import { MINLENGTH } from './fuzzy-input.consts';

//--- View Funktionen -----

export function isValid(input: string, props: Props): boolean {
    if((typeof props.valid !== 'undefined') && !props.valid) {
        return false;
    } else if(props.pattern) {
        return props.pattern.test(input);
    } else {
        return true;
    }
}

export function isReady(input: string, currentValue: string, state: State, props: Props): boolean {
    const minLength = (props && typeof props.minLength === 'number') 
        ? props.minLength 
        : MINLENGTH;
    return input.length >= minLength
        && input === currentValue
        && !state.loading;
}

export function findMatch(input: string, prefix: string): string|null {
    // Prefix gefolgt von erlaubten Zeichen, endend mit Leerzeichen oder EOL.
    const pattern = `${prefix}[0-9a-zA-ZüöäÜÖÄß_#%\\?\\-\\+]*(\\s|$)`;
    const regex = new RegExp(pattern, 'gmi');
    return input.match(regex)?.[0]?.trim() ?? null;
}
