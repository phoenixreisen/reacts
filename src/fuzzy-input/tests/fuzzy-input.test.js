/**
 * @jest-environment jsdom
 */
import { render, fireEvent, waitFor, screen, getByText } from '@testing-library/react';
import shallowRenderer from 'react-test-renderer/shallow';
import { isReady, isValid } from '../fuzzy-input.fns';
import renderer, { act } from 'react-test-renderer';
import FuzzyInput from '../fuzzy-input.r';
import React from 'react';

// Erweiterung von expect() um DOM-Funktionen
import '@testing-library/jest-dom';

describe('Fuzzy Input', () => {
    jest.useFakeTimers();

    test('should render correctly oninit', () => {
        const component = renderer.create(
            <FuzzyInput
                throttling={5}
                onQuery={() => Promise.resolve(['S1', 'S2', 'S3'])}
                onLoad={() => Promise.resolve({ type: 'success', value: 'String 2'})}
            />
        );
        const snap = component.toJSON();
        expect(snap).toMatchSnapshot();
    });

    test('should render button correctly', () => {
        const withoutButton = renderer.create(
            <FuzzyInput
                inText={undefined}
                withButton={false}
            />
        );
        const snapWithoutButton = withoutButton.toJSON();
        expect(snapWithoutButton).toMatchSnapshot();

        const withButton = renderer.create(
            <FuzzyInput
                withButton={true}
                inText={{ prefix: '{', suffix: '}' }}
            />
        );
        const snapWithButton = withButton.toJSON();
        expect(snapWithButton).toMatchSnapshot();
    });

    test('should correctly set and use input value from inside or outside', () => {
        const c = renderer.create(
            <FuzzyInput
                onInput={undefined}
            />
        );
        expect(c.root.findByType('input').props.value).toEqual('');
        act(() => c.root.findByType('input').props.onInput({ target: { value: 'test' } }));
        expect(c.root.findByType('input').props.value).toEqual('test');

        const outerState = {
            value: ''
        };
        const cOut = renderer.create(
            <FuzzyInput
                value={outerState.value}
                onInput={(value) => outerState.value = `${value} from outer space`}
            />
        );
        expect(cOut.root.findByType('input').props.value).toEqual('');
        act(() => cOut.root.findByType('input').props.onInput({ target: { value: 'test' } }));
        // Simulated re-render of the outer component, causing the inner 
        // component to be re-rendered with a new value provided from the outside
        cOut.update(
            <FuzzyInput
                value={outerState.value}
                onInput={(value) => outerState.value = `${value} from outer space`}
            />
        );
        expect(cOut.root.findByType('input').props.value).toEqual('test from outer space');
    });

    test('should correctly return input value', () => {
        const cInnter = renderer.create(
            <FuzzyInput />
        )
        expect(cInnter.root.findByType('input').props.value).toEqual('');
        act(() => cInnter.root.findByType('input').props.onInput({ target: { value: 'inner value' } }));
        expect(cInnter.root.findByType('input').props.value).toEqual('inner value');

        const cOuter = renderer.create(
            <FuzzyInput
                value={'outer value'}
            />
        );
        expect(cOuter.root.findByType('input').props.value).toEqual('outer value');
        act(() => cOuter.root.findByType('input').props.onInput({ target: { value: 'inner value' } }));
        // Der Feldinput wird immer über die äußere Property bezogen. Ohne eine onInput-Funktion
        // von außen, bleiben Änderungen am Feld ohne Wirkung.
        expect(cOuter.root.findByType('input').props.value).toEqual('outer value');
    });

    test('should correctly validate input', () => {
        const attrsMock = { valid: true, pattern: undefined };
        expect(isValid('test', attrsMock)).toBe(true);
        attrsMock.valid = false;
        expect(isValid('test', attrsMock)).toBe(false);
        attrsMock.valid = true;
        attrsMock.pattern = new RegExp(/^[0-9]+$/);
        expect(isValid('123', attrsMock)).toBe(true);
        expect(isValid('test', attrsMock)).toBe(false);
        expect(isValid('test123', attrsMock)).toBe(false);
        attrsMock.valid = undefined;
        attrsMock.pattern = undefined;
        expect(isValid('test', attrsMock)).toBe(true);
    });

    test('should correctly determine if input is ready for search', () => {
        const state = { loading: false, value: 'test' };
        const attrs = { minLength: undefined };
        let input = 'test';
        expect(isReady(input, input, state, attrs)).toBe(true);
        state.loading = true;
        expect(isReady(input, input, state, attrs)).toBe(false);
        state.loading = false;
        state.value = 'something other';
        expect(isReady(input, state.value, state, attrs)).toBe(false);
        input = 'te';
        state.value = 'te';
        // default min length ist aber 3 Zeichen
        expect(isReady(input, input, state, attrs)).toBe(false);
        attrs.minLength = 2;
        expect(isReady(input, input, state, attrs)).toBe(true);
    });

    test('should correctly focus a certain DOM item by Id', async () => {
        let searchSpyCount = 0;
        
        const up = { key: 'ArrowUp', preventDefault: jest.fn() };
        const down = { key: 'ArrowDown', preventDefault: jest.fn() };
        const searchSpy = jest.spyOn(window, 'setTimeout');

        const comp = renderer.create(
            <FuzzyInput
                onQuery={() => Promise.resolve(['Test 1', 'Test 2', 'Test 3'])}
                onLoad={() => Promise.resolve({ type: 'success', value: 'Test 3 ausgewählt'})}
            />
        );
        const rootElement = comp.root.findByProps({
            className: 'fuzzy-search fuzzy-show-result'
        });

        const tree = comp.toJSON();
        // console.log('TREE', tree);
        expect(tree).toMatchSnapshot();
        expect(rootElement).not.toBeNull();
        expect(searchSpy).toHaveBeenCalledTimes(0);
        
        act(() => {
            comp.root.findByType('input').props.onInput({ target: { value: 'Test' } })
        });
        expect(searchSpy).toHaveBeenCalledTimes(1);
        expect(comp.root.findByType('input').props.value).toEqual('Test');

        await waitFor(() => {
            const overlay = comp.root.findByProps({
                className: 'fuzzy-overlay fuzzy-result fuzzy-style'
            });
            expect(overlay).not.toBeNull();
            const treeWithOverlay = comp.toJSON();
            expect(treeWithOverlay).toMatchSnapshot();
        });
    });

    test('should call query and search callback in autocomplete mode correctly', async () => {
        const search = 'Result';
        const resultList = [
            'Result 1',
            'Result 2',
            'Result 3',
            'Blabla 1',
            'Blabla 2',
            'Anything 1',
            'Anything 2',
        ];
        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                onQuery={() => Promise.resolve(resultList.filter((item) => 
                    item.includes(search)
                ))}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: search } });
        expect(input.value).toBe(search);

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();

            const resultDiv = queryByTestId('fuzzy-result');
            const children = resultDiv.querySelectorAll('a');
            expect(children.length).toBe(3);

            expect(getByText('Result 1')).toBeInTheDocument();
            expect(getByText('Result 2')).toBeInTheDocument();
            expect(getByText('Result 3')).toBeInTheDocument();
        });
    });

    test('should call query and search callback in placeholder mode correctly', async () => {
        const resultList = [
            'Result 1',
            'Result 2',
            'Result 3',
            'Blabla 1',
            'Blabla 2',
            'Anything 1',
            'Anything 2',
        ];
        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                inText={{ prefix: '{{', suffix: '}}' }}
                onQuery={() => Promise.resolve(resultList)}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: '{{' } });
        expect(input.value).toBe('{{');

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();

            const resultDiv = queryByTestId('fuzzy-result');
            const children = resultDiv.querySelectorAll('a');
            expect(children.length).toBe(7);

            expect(getByText('Result 1')).toBeInTheDocument();
            expect(getByText('Blabla 1')).toBeInTheDocument();
            expect(getByText('Anything 1')).toBeInTheDocument();
        });
    });

    test('should provide up and down going (focusing) within result list by arrow keys', async () => {
        const { getByRole, getByText, getByTestId } = render(
            <FuzzyInput
                onQuery={() => Promise.resolve(['Test 1', 'Test 2', 'Test 3'])}
                onLoad={() => Promise.resolve({ type: 'success', value: 'Test 3 ausgewählt'})}
            />
        );
    
        // Type test into the input field
        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: 'Test' } });
        expect(input.value).toBe('Test');
    
        // Wait till the result list is shown
        await waitFor(() => {
            expect(getByTestId('fuzzy-result')).toBeInTheDocument();
        });
    
        // Press arrow down key
        fireEvent.keyDown(document.body, { key: 'ArrowDown', code: 'ArrowDown' });

        // Check if the first element has the focus
        await waitFor(() => {
            const firstResultItem = getByTestId('fuzzy-input-item-0');
            expect(firstResultItem).toHaveFocus();
        });

        // Press arrow down key
        fireEvent.keyDown(document.body, { key: 'ArrowDown', code: 'ArrowDown' });

        // Check if the second element has the focus
        await waitFor(() => {
            const secondResultItem = getByTestId('fuzzy-input-item-1');
            expect(secondResultItem).toHaveFocus();
        });

        // Press arrow up key
        fireEvent.keyDown(document.body, { key: 'ArrowUp', code: 'ArrowUp' });

        // Check if the first element has the focus
        await waitFor(() => {
            const firstResultItem = getByTestId('fuzzy-input-item-0');
            expect(firstResultItem).toHaveFocus();
        });
    });

    test('should close result list when escape (ESC) key is pressed', async () => {
        const { getByRole, getByText, queryByTestId } = render(
            <FuzzyInput
                onQuery={() => Promise.resolve(['Test 1', 'Test 2', 'Test 3'])}
                onLoad={() => Promise.resolve({ type: 'success', value: 'Test 3 ausgewählt'})}
            />
        );

        // Type test into the input field
        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: 'Test' } });
        expect(input.value).toBe('Test');

        // Wait till the result list is shown
        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();
        });

        // Close/remove result list by pressing escape key
        fireEvent.keyUp(document.body, { key: 'Escape', code: 'Escape' });

        // Check whether the result list is removed from DOM
        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).not.toBeInTheDocument();
        });
    });

    test('should correctly find placeholder matches', async () => {
        const resultList = [
            'Mein Name ist Mia', 
            'Der Hund ist tot', 
            'Der Hund wird tot sein',
            'Der Hund wird tot gewesen sein'
        ];

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                inText={{
                    prefix: '{{',
                    suffix: '}}'
                }}
                onQuery={() => Promise.resolve(resultList)}
                onLoad={() => Promise.resolve({
                    type: 'success',
                    value: resultList[3]
                })}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: 'Test' } });
        expect(input.value).toBe('Test');

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).not.toBeInTheDocument();
        });

        fireEvent.input(input, { target: { value: 'Sam sagt: {{Hu' } });

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();
            expect(getByText('Der Hund wird tot gewesen sein')).toBeInTheDocument();
            // Three items should be in the result list, those with "Hund" in it
            expect(getByTestId('fuzzy-input-item-0')).toBeInTheDocument();
            expect(getByTestId('fuzzy-input-item-1')).toBeInTheDocument();
            expect(getByTestId('fuzzy-input-item-2')).toBeInTheDocument();
            expect(queryByTestId('fuzzy-input-item-3')).not.toBeInTheDocument();
        });

        fireEvent.keyDown(document.body, { key: 'ArrowDown', code: 'ArrowDown' });
        await waitFor(() => {
            const result1 = getByTestId('fuzzy-input-item-0');
            expect(result1).toHaveTextContent('Der Hund ist tot');
            expect(result1).toHaveFocus();
        });

        fireEvent.keyDown(document.body, { key: 'ArrowDown', code: 'ArrowDown' });
        await waitFor(() => {
            const result2 = getByTestId('fuzzy-input-item-1');
            expect(result2).toHaveTextContent('Der Hund wird tot sein');
            expect(result2).toHaveFocus();
        });

        fireEvent.keyDown(document.body, { key: 'ArrowDown', code: 'ArrowDown' });
        await waitFor(() => {
            const result3 = getByTestId('fuzzy-input-item-2');
            expect(result3).toHaveTextContent('Der Hund wird tot gewesen sein');
            expect(result3).toHaveFocus();
        });

        const result3 = getByTestId('fuzzy-input-item-2');
        fireEvent.keyUp(result3, { key: 'Enter', code: 'Enter', charCode: 13 });
        await waitFor(() => {
            const input = getByRole('textbox');
            expect(input.value).toBe(`Sam sagt: {{${resultList[3]}}} `);
        });
    });

    test('should show an error when a promise rejection occurs', async () => {
        const error = async () => Promise.reject({type: 'failure', msg: 'bla'});
        const needle = 'Test';

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                onQuery={() => Promise.reject(error)}
            />
        );
        
        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(document.querySelector('.fuzzy-error')).toBeNull();
        expect(input.value).toBe(needle);

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).not.toBeInTheDocument();
            expect(document.querySelector('.fuzzy-error')).toBeInTheDocument();
            expect(getByText('Oha, während der Abfrage ist ein Fehler aufgetreten.')).toBeInTheDocument();
        });
    });

    test('should call onLoad callback correctly', async () => {
        const onLoadFn = jest.fn();
        const needle = 'Test';

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                onQuery={() => Promise.resolve([`${needle} 1`, `${needle} 2`, `${needle} 3`])}
                onLoad={() => onLoadFn()}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(input.value).toBe(needle);

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();
        });

        const result1 = getByTestId('fuzzy-input-item-0');
        fireEvent.keyUp(result1, { key: 'Enter', code: 'Enter', charCode: 13 });

        await waitFor(() => {
            expect(onLoadFn).toHaveBeenCalledTimes(1);
            expect(getByRole('textbox').value).toBe(`${needle} 1`);
        });
    });

    test('should clear field, when clearAfterLoad is set', async () => {
        const onLoadFn = jest.fn();
        const needle = 'Test';

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                clearAfterLoad={true}
                onLoad={() => onLoadFn()}
                onQuery={() => Promise.resolve([`${needle} 1`, `${needle} 2`, `${needle} 3`])}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(input.value).toBe(needle);

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();
        });

        const result1 = getByTestId('fuzzy-input-item-0');
        fireEvent.keyUp(result1, { key: 'Enter', code: 'Enter', charCode: 13 });
        
        await waitFor(() => {
            expect(getByRole('textbox').value).toBe('');
            expect(onLoadFn).toHaveBeenCalledTimes(1);
        });
    });

    test('should show warning, when input is invalid', async () => {
        const onLoadFn = jest.fn();
        const needle = 'Test';

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                clearAfterLoad={true}
                pattern={new RegExp(/[0-9]/)}
                onQuery={() => Promise.resolve([`${needle} 1`, `${needle} 2`, `${needle} 3`])}
                onLoad={() => onLoadFn()}
            />
        );
        
        await waitFor(() => {
            expect(document.querySelector('.fuzzy-warning')).not.toBeInTheDocument();
        });

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(input.value).toBe(needle);

        await waitFor(() => {
            expect(document.querySelector('.fuzzy-warning')).toBeInTheDocument();
            expect(getByText('Ungültige Eingabe.')).toBeInTheDocument();
        });
    });

    test('should catch error, when onLoad failed', async () => {
        const spy = jest
            .spyOn(console, 'error')
            .mockImplementation(() => {});

        const mockFunction = jest.fn(() => {
            throw new Error('Failure');
        });
        const needle = 'Test';

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                onQuery={() => Promise.resolve([`${needle} 1`, `${needle} 2`, `${needle} 3`])}
                onLoad={() => mockFunction()}
                clearAfterLoad={true}
                logerror={true}
            />
        );

        expect(document.querySelector('.fuzzy-error')).not.toBeInTheDocument();

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(input.value).toBe(needle);

        await waitFor(() => {
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();
        });

        const result1 = getByTestId('fuzzy-input-item-0');
        fireEvent.keyUp(result1, { key: 'Enter', code: 'Enter', charCode: 13 });

        await waitFor(() => {
            expect(spy).toHaveBeenCalled();
            expect(mockFunction).toHaveBeenCalled();
            expect(document.querySelector('.fuzzy-error')).not.toBeInTheDocument();
        });
    });

    test('should blur if onBlur is set', async () => {
        const needle = 'Test';
        const mockFunction = jest.fn();

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                onQuery={() => Promise.resolve([`${needle} 1`, `${needle} 2`, `${needle} 3`])}
                onLoad={() => fireEvent.input(getByRole('textbox'), { target: { value: 'loaded' } })}
                onBlur={() => mockFunction()}
                clearAfterLoad={true}
                logerror={true}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(input.value).toBe(needle);

        fireEvent.blur(input);

        await waitFor(() => {
            expect(mockFunction).toHaveBeenCalled();
        });
    });

    test('should trottle search requests correctly - two times', async () => {
        jest.useFakeTimers();

        const mockFunction = jest.fn().mockImplementation(() => {
            return Promise.resolve([`${needle} 1`, `${needle} 2`, `${needle} 3`])
        });
        const needle = 'Test';

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                throttling={300}
                onQuery={() => mockFunction()}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(input.value).toBe(needle);

        jest.advanceTimersByTime(310);

        fireEvent.input(input, { target: { value: `${needle} 1` } });
        expect(input.value).toBe(`${needle} 1`);

        await waitFor(() => {
            // A throttling duration of 300ms is set.
            // We fill the form two times with a pause of 310ms.
            // So the function should be called two times
            expect(mockFunction).toHaveBeenCalledTimes(2);
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();
        });

        jest.useRealTimers();
    });

    test('should trottle search requests correctly - one time', async () => {
        jest.useFakeTimers();

        const mockFunction = jest.fn().mockImplementation(() => {
            return Promise.resolve([`${needle} 1`, `${needle} 2`, `${needle} 3`])
        });
        const needle = 'Test';

        const { getByRole, getByText, getByTestId, queryByTestId } = render(
            <FuzzyInput
                throttling={300}
                onQuery={() => mockFunction()}
            />
        );

        const input = getByRole('textbox');
        fireEvent.input(input, { target: { value: needle } });
        expect(input.value).toBe(needle);

        jest.advanceTimersByTime(200);

        fireEvent.input(input, { target: { value: `${needle} 1` } });
        expect(input.value).toBe(`${needle} 1`);

        await waitFor(() => {
            // A throttling duration of 300ms is set.
            // We fill the form two times with a pause of 200ms.
            // So the function should be called one time with the second input, 
            // because it's not yet throttled
            expect(mockFunction).toHaveBeenCalledTimes(1);
            expect(queryByTestId('fuzzy-result')).toBeInTheDocument();
        });

        jest.useRealTimers();
    });
});