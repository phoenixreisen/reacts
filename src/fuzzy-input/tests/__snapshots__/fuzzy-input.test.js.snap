// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fuzzy Input should correctly focus a certain DOM item by Id 1`] = `
<article
  className="fuzzy-search fuzzy-show-result"
>
  <div
    className=""
  >
    <label
      className="textfield fuzzy-input"
    >
      <input
        autoComplete="off"
        data-testid="fuzzy-input"
        disabled={false}
        id="fuzzy-input"
        maxLength={100}
        name="fuzzy-input"
        onInput={[Function]}
        placeholder="..."
        readOnly={false}
        type="text"
        value=""
      />
      <span
        className="textfield__label"
      >
        Suche
      </span>
    </label>
  </div>
</article>
`;

exports[`Fuzzy Input should correctly focus a certain DOM item by Id 2`] = `
<article
  className="fuzzy-search fuzzy-show-result"
>
  <div
    className=""
  >
    <label
      className="textfield fuzzy-input"
    >
      <input
        autoComplete="off"
        data-testid="fuzzy-input"
        disabled={false}
        id="fuzzy-input"
        maxLength={100}
        name="fuzzy-input"
        onInput={[Function]}
        placeholder="..."
        readOnly={false}
        type="text"
        value="Test"
      />
      <span
        className="textfield__label"
      >
        Suche
      </span>
    </label>
  </div>
  <div
    className="fuzzy-overlay fuzzy-result fuzzy-style"
    data-testid="fuzzy-result"
  >
    <div
      className="fadeIn animated faster"
    >
      <a
        data-testid="fuzzy-input-item-0"
        href="#"
        id="fuzzy-input-item-0"
        onClick={[Function]}
        onKeyUp={[Function]}
      >
        Test 1
      </a>
      <a
        data-testid="fuzzy-input-item-1"
        href="#"
        id="fuzzy-input-item-1"
        onClick={[Function]}
        onKeyUp={[Function]}
      >
        Test 2
      </a>
      <a
        data-testid="fuzzy-input-item-2"
        href="#"
        id="fuzzy-input-item-2"
        onClick={[Function]}
        onKeyUp={[Function]}
      >
        Test 3
      </a>
    </div>
  </div>
  <div
    className="fuzzy-bg-layer"
    onClick={[Function]}
  />
</article>
`;

exports[`Fuzzy Input should render button correctly 1`] = `
<article
  className="fuzzy-search fuzzy-show-result"
>
  <div
    className=""
  >
    <label
      className="textfield fuzzy-input"
    >
      <input
        autoComplete="off"
        data-testid="fuzzy-input"
        disabled={false}
        id="fuzzy-input"
        maxLength={100}
        name="fuzzy-input"
        onInput={[Function]}
        placeholder="..."
        readOnly={false}
        type="text"
        value=""
      />
      <span
        className="textfield__label"
      >
        Suche
      </span>
    </label>
  </div>
</article>
`;

exports[`Fuzzy Input should render button correctly 2`] = `
<article
  className="fuzzy-search fuzzy-show-result"
>
  <div
    className="fuzzy-with-button"
  >
    <label
      className="textfield fuzzy-input"
    >
      <input
        autoComplete="off"
        data-testid="fuzzy-input"
        disabled={false}
        id="fuzzy-input"
        maxLength={100}
        name="fuzzy-input"
        onInput={[Function]}
        placeholder="..."
        readOnly={false}
        type="text"
        value=""
      />
      <span
        className="textfield__label"
      >
        Suche
      </span>
    </label>
    <button
      className="btn btn--secondary"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-list"
      />
    </button>
  </div>
</article>
`;

exports[`Fuzzy Input should render correctly oninit 1`] = `
<article
  className="fuzzy-search fuzzy-show-result"
>
  <div
    className=""
  >
    <label
      className="textfield fuzzy-input"
    >
      <input
        autoComplete="off"
        data-testid="fuzzy-input"
        disabled={false}
        id="fuzzy-input"
        maxLength={100}
        name="fuzzy-input"
        onInput={[Function]}
        placeholder="..."
        readOnly={false}
        type="text"
        value=""
      />
      <span
        className="textfield__label"
      >
        Suche
      </span>
    </label>
  </div>
</article>
`;
